### ISSUE TITLE
Using the compat distro kernel

### ISSUE THREAD

**Issue by wdlkmpx** [puppylinux-woof-CE/woof-CE#1372]
In woofce it's possible to use the compat distro kernel in initrd.gz & zdrv, well theoretically. 

Sometime ago, i was able to boot a debian kernel, but i think it wasn't the default kernel, that kernel had basically everything as .ko modules. It was confusing at the time, and it still is for me.

I added 2 helper scripts (modded stuff from the old 3builddistro), and a helper script to the initrd (from the old init). 

I booted to desktop using gyro's early overlayfs stuff, but it was long ago, i've kind of forgotten what i did, but the code is there. And i noticed a "high" cpu usage.. too many modules loaded in the initrd.

It's really good to have the ability to compile any kernel version with the kernel-kit, but let's not forget that you have to recompile stuff for your kernel, including proprietary drivers.

The big distros include proprietary video drivers and stuff ready to use, but i think they're not usable in puppy, so one of the big advantages is lost, you have to put on your robe and wizard hat. Although without this severe limitations, people wouldn't know how to overcome this.

This is a complex issue that those who never use the big distros themselves don't understand (including myself).

The kernel i used was probably not the correct one, and the method i used was probably not the correct one.

Toni or Saintless shared this knowledge some time ago:

> I think I have what I want already. My goal is to make possible to boot any existing Puppy
with official Debian/Ubuntu kernel. Not like I really need it since I don't use Puppy except for testing.

This is what I have as a result:
https://github.com/MintPup/Puppy-Linux/tree/master/Debian-kernel

I will explain the steps in the next few days.

After including some kernel drivers in initrd.gz from official debian initrd.img (generated with live-boot) - this is the only change inside init script to make it work:
https://github.com/MintPup/Puppy-Linux/commit/d3dc7735692830e9ed9590893aff7a2efb9bbac9

Save file (encrypted on not), first save, save folder and all tested works with current Puppy build like Puduan-6.0.0 from attentionseeker0
For older puppy like saluki and dpup it is the same except save folder (it wasn't available at that time as option and I guess it needs changes in most scripts). But those old pups do not have option to select save folder anyway at first shutdown so it isn't much of a problem.

**Comment by mavrothal** [comment]
With the current scheme of zdrv and easy kernel change, you need certain modules builtin. Aufs (or maybe overlay in the future) and squashfs. Most distro kernels have them as modules. 
Given the rate that distros update their kernel, you certainly need the easy change to keep up.
You can use distro kernels going back to the old style puppy initrd or something similar, where certain modules where in the initrd and used to setup that layered system before chrooting, but then easy kernel change is not going to be very easy, unless you use distro initrd too.
Another concern is that Aufs in distro kernels is usually outdated, so unless you want to drop aufs altogether you may get a kernel from a distro that keeps up with aufs (do not know which other than puppy)
Another thing to consider is that if you use everything from a distro, from the initrd image and the kernel to package manager to toolchain to apps to systemd (some distro kernels might need it), to whatever, adding few extra utility scripts can hardly make you a "distro" 😸 
Going back to the main reason, proprietary drivers, is usually 2-3 out of tree drivers needed (nvidia, broadcom, what else?). Adding a little code in kernel kit to compile those too, might not be too difficult.
Anyway, you do what you do!

**Comment by wdlkmpx** [comment]
I think the kernel i used didn't have aufs. That's why i used gyro's overlayfs init. And i think it wasn't the default debian kernel.

It downloaded the .deb kernel, created the zdrv and added a few drivers to the initrd.

I think the distro kernel should be specified in the PKGS_SPECS, and it should be ignored/deleted if a huge_kernel is selected or specified in build.conf, etc.. that way the logic is simplified.

I see woodenshoe-wi added code here and there, i guess i should buy a raspberry pi to test that stuff too and see if i can unify stuff.

I'm also tempted to delete the code i added, i'm more inclined to delete it actually.

**Comment by woodenshoe-wi** [comment]
If I remember correctly, when I was still using the Raspbian kernels the Raspup images produced were structured similarly to the official Raspbian images.  A small FAT partition with the kernels and firmware followed by an ext root partition.  It was basically a full install without any aufs or overlayfs.

After I got the kernel-kit to make ARM kernels with aufs I switched the Raspup build to making zip files with sfs files in them.  I think sfs support was a separate kernel module in the Raspbian kernels, so keeping all the kernel modules in a zdrv sfs wouldn’t work without putting a copy of the squashfs kernel module somewhere else.

Because I think I have a problem with my PC’s hard drive I don’t want to do any woof builds or compiling for PC’s, but if you need something compiled on a raspberry pi I can still help.  I also have an extra SD card and can help test any finished builds if needed.

**Comment by wdlkmpx** [comment]
I realize that using the compat distro kernel is basically the only way to get proper support in some arm boards. In the powerful arm boards. There is a way to create zdrv (kernel_pkg.sh) and the initrd.gz (kernel_pkg_zdrv2initrd.sh), but it's flawed, experimenting with it might yield some results, but all the energy has been drained out of me.

Seeing some "fairly" local prices:
pi zero: $30
pi 1: $30
pi2: $45
pi3: $60

Does not include shipping

I want a pi3 model b+... aarch64 will be added.

woodenshoe-wi, well, i'd like to see builds of raspup. You might want to create a dedicated page or a git wiki page where you explain how to install it and provide a build for testing purposes.. that will be deleted once a new build is available. A rolling release raspup, since fixing puppy will take a lifetime. With a big disclaimer: "i do this in good faith, and no one has the right to complain". so you can find a peace of mind.

I think there should be an option to build a raspup for the pi zero, pi 1, pi2 & p3

You might want to explain how you would implement that, and then i'll tell you how i would do that, of course in the experimental branch. wmsgs is available if you want some privacy.

Due to limited ram and using a flash storage device, the woofce configs for the pi boards should be minimal, a minimal gui and cli-only for pi zero & pi 1.

I don't know maybe it should be where we make puppy completely compatible with busybox applets by omitting coreutils, util-linux, of course testing that it works with both bb and the full apps.

This is an example i'll add later... all references to disktype will be removed, another puppy app that will be gone.

```shell
WOOF_OUTPUT=.
PUPIMG=raspi-sd-1gb-skeleton-ext4_nojournal-8nov2016.img
..
.....
......

b() {
	BB=/bin/busybox
	#SDIMGINFO="`LANG=C $BB fdisk -l ../${WOOF_OUTPUT}/${PUPIMG}`"
	SDIMGINFO="`LANG=C $BB fdisk -l ${WOOF_OUTPUT}/${PUPIMG}`"
	IS_EXT2="`echo "$SDIMGINFO" | grep 'Linux'`"
	if ! [ "$IS_EXT2" ] ; then
		echo -n "need ext4 in the second partition of ${SDIMAGE}, aborting."
		exit 1
	fi
	# Units: sectors of 1 * 512 = 512 bytes
	BYTESPERSECTOR="`echo "$SDIMGINFO" | grep 'sectors of' | cut -f 2 -d '=' | cut -f 2 -d ' '`"
	p1=`echo "$SDIMGINFO" | grep "${PUPIMG}1" | tr -s ' '`
	p2=`echo "$SDIMGINFO" | grep "${PUPIMG}2" | tr -s ' '`
	if [ "`echo "$SDIMGINFO" | grep 'StartCHS'`" ] ; then
		#busybox
		if [ "`echo "$p1" | grep '\*'`" ] ; then
			read IMG BOOT STARTCHS ENDCHS P1STARTSECTORS P1ENDSECTOR P1SECTORS ETC <<< "$p1"
		else
			read IMG STARTCHS ENDCHS P1STARTSECTORS P1ENDSECTOR P1SECTORS ETC <<< "$p1"
		fi
		read IMG STARTCHS ENDCHS P2STARTSECTORS P2ENDSECTOR P2SECTORS ETC <<< "$p2"
	else
		if [ "`echo "$p1" | grep '\*'`" ] ; then
			read IMG BOOT P1STARTSECTORS P1ENDSECTOR P1SECTORS ETC <<< "$p1"
		else
			read IMG P1STARTSECTORS P1ENDSECTOR P1SECTORS ETC <<< "$p1"
		fi
		read IMG P2STARTSECTORS P2ENDSECTOR P2SECTORS ETC <<< "$p2"
	fi
	P1BYTES=$((P1SECTORS * BYTESPERSECTOR))
	P1STARTBYTES=$((P1STARTSECTORS * BYTESPERSECTOR))
	P2BYTES=$((P2SECTORS * BYTESPERSECTOR))
	P2STARTBYTES=$((P2STARTSECTORS * BYTESPERSECTOR))
	echo;echo
	echo "P1BYTES=$P1BYTES"
	echo "P1SECTORS=$P1SECTORS"
	echo "P1STARTSECTORS=$P1STARTSECTORS"
	echo "BYTESPERSECTOR=$BYTESPERSECTOR"
	echo "P1STARTBYTES=$P1STARTBYTES"
	echo
	echo "P2BYTES=$P2BYTES"
	echo "P2SECTORS=$P2SECTORS"
	echo "P2STARTSECTORS=$P2STARTSECTORS"
	echo "P2STARTBYTES=$P2STARTBYTES"
	echo
	#echo $p1
	#echo $p2
}

b
```

**Comment by woodenshoe-wi** [comment]
The Raspberry Pis seem to run fine with the kernels I made using kernel-kit.  I had to modify kernel-kit to support cross compiling ARM kernels, but those changes have already been merged.

> I think there should be an option to build a raspup for the pi zero, pi 1, pi2 & p3

Already done 😄
One build will work on any of the Pis.  Raspup is too fat to load the main sfs in RAM on the Pi zero (it runs from the SD card instead), but it will run in RAM on any of the Pis that have 1G of RAM.

The computer I was using to compile stuff and do woof builds is out of order right now, and with spring time coming I have yard work and gardening to keep up with too.

Here in the US a Pi3+ is available for $35 USD plus maybe $10 or $12 USD shipping,
https://www.adafruit.com/product/3775
But I’m afraid that shipping stuff out of the US is prohibitively expensive.  My dad mailed a box to his brothers in the Netherlands and it was over $50.

I did add some information to the readme at https://github.com/woodenshoe-wi/woof-CE
Everything in that branch should already be merged into testing, except the charges to the readme file of course.  It is out of date, so any builds should be made with the puppylinux-woof-CE/woof-CE testing branch instead.

 I can update the readme, should I direct people back to the puppylinux-woof-CE/woof-CE page or did you want discussion about Raspberry Pis to be kept at woodenshoe-wi/woof-CE?  I don’t have time to keep a build up to date with the current testing branch even if my computer wasn’t out of order.

**Comment by mavrothal** [comment]
The problem with ARM boards is that each one (family) needs its own kernel. Using a distro kernel will just solve the issue for a specific board. On the other hand modifying a distro kernel to work with puppy is fairly simple. All you need is aufs and a [config patch](https://github.com/mavrothal/Pox_git/blob/master/XO/XO_kernel_patches/3.5arm/3.5_aufs_xo_4_defconfig.patch). Having the current puppy kernel setup/init you can just take kernels from any distro that boot a specific board modify them and you are good to go.See the [ArmDog thread](http://murga-linux.com/puppy/viewtopic.php?t=92548) for more. See also a [howto from James](http://lightofdawn.org/wiki/wiki.cgi/FatdogArm/AdoptingFatdogArm).
Short story, would be more effective to have an ```armedpuppy.sfs``` and an ```armedinird.img``` and compile ```board-specific_zdrv.sfs/vmlinuz``` based on kernels that boot the said boards. 
Then provide either multiple board-specific downloads or one downloads with multiple ZDRVs and a menu to boot the specified board (fancy but a waste I think unless the kernels are really small 10-15MB).

**Comment by wdlkmpx** [comment]
I think the readme should be in woof-CE/woof-distro/arm/raspbian or a wiki page in the repo or something, well the discussion can happen anywhere when there is something to say. We should also write random docs in a new dir: docs/

I read the the readme, the buildsys produces a zip file which must extracted into a FAT formatted SD.

Regarding other boards, hmm well, builders with the hardware are needed to see what can be done. I like the idea of profiles.

There is someone in my city selling a used pi1 including hdmi cable, sd card? and kodi? ready to watch HD videos? O_o? for $15.. it must be about to die.. I'm trying to contact him.

After some testing i'll comment about it, i can build stuff for testing (using the experimental branch) using old hard drives and and old pc, it's incredible how much i abused those old hard drives compiling wine, building isos, and they don't die, the same speed remains.

Major speedups in the build system thanks to those old hard drives and pc's, because when i'm building with a fast machine i hardly notice any difference.

**Comment by wdlkmpx** [comment]
raspi 1 a2 or b2 + 8gb sd with openelec + hdmi cable + [2] 16gb micro sd's (with sd adapter) + 2 usb hubs (4 ports) = $24

I'm tempted to buy a raspi 3 b+, $60 unfortunately. I didn't think it was soooo small.
 
Now that i have a raspi i built a raspup, i'm yet to test it.
https://github.com/puppylinux-woof-CE/woof-CE/wiki/Builds

It's a little big for a raspi, i'm sure some pkgs can be removed (llvm, etc).

**Comment by woodenshoe-wi** [comment]
3 SD cards, a hdmi cable and 2 usb hubs sounds like a good deal even if you do decide to get the pi3 later.  The 2 micro SD cards can be used with the newer Pis, you can never have too many SD cards 😄.

I always compile stuff on my pi1 even though it is the slowest one because it has the same chip as the pi zero.  This way the configure scripts are forced to compile the program so that it will run on any of the Pis.  I think the official Raspbian builds are done on more powerful ARM devices, but they have something that scans the binaries to see if there are any illegal opcodes for the smaller chips.

If you want to use a GUI/X environment, the quad core Pis (pi2, pi3) are noticeably more responsive, but they are still much slower than modern computers.

The reason I say “modern computers” is that when I was a child my parents brought a 286 pc that ran MS-DOS.  I was fortunate enough to grow up with a computer that I could learn about.  MS-DOS came with a BASIC interpreter, and when I saw my dad programming, I practically taught myself programming from the help documentation of quick-basic 4.5.

As people seem to be moving to iOS and Android devices and away from “real” computers, I’m hoping that the Raspberry Pis and Python interpreter can give children the opportunity to experiment the way I did with a 286 and quick-basic.  It doesn’t necessarily need a GUI, but needs to be simple to understand.  There are apps that let you connect to computers using ssh from your phone or tablet, so it should be possible to use a Pi that has wifi without even connecting to a TV and keyboard.  On the other hand, if the goal is to get the children off the phone/tablet, maybe connecting to the TV like video games used to before there even were “personal computers” would be a better idea.

**Comment by wdlkmpx** [comment]
I was not as fortunate, i learned a bit of basic when i was 20 i
think. But some time later i found that writing mIRC scripts was cool,
after years the next thing i tried was shell scripting and a bit of C.
And that's the only thing i need in a linux system, a shell
interpreter + tools and a C compiler. True minimalism, that's i think
i'll like the pi's.

The guy who sold me the raspi said he's going to buy a 3d printer and
will print a raspi case for me ($0). This one looks nice:
https://www.thingiverse.com/thing:25100

I'll buy a printer + continuous ink system for my sister tomorrow, so
i guess i'll buy the raspi 3 next month when the money comes back.

I also want this: https://www.hardkernel.com/shop/odroid-n2-with-4gbyte-ram/

**Comment by wdlkmpx** [comment]
I just tested raspup stretch in my raspi 1 b

For some reason some apps don't run (urxvt, retrovol,etc).. not found,
but file identifies them as elf binaries. Must be an issue with the sd
card or ram or aufs?. I don't see aufs errors in dmesg.

Jwm needs some themes, i'll adjust stuff.

rox is quite usable with pcmanfm keybindings. I see i can run the
terminal with 2 keys: ` and F4

I wonder if it's possible to:
Trigger 'Display->List View' with Ctrl-4 (and optionally with Ctrl-2)
Trigger 'Display->Icons View' with Ctrl-3  (and optionally with Ctrl-1)

Continuing.

After rebooting a few times, i decided to create a savefile.. it
apparently can't create a ext4 savefile in a fat32 partition (sd
card).  I'll try with ext3 later.

But overall i find this amazing , it just shows how much puppy has
been optimized. And an excellent job by woodenshoe-wi

It's also amazing how small this boards are. There are also banana
pi's, orange pi's but what if a want a watermelon pi.

I was about test the init bootmenu, but it froze while trying to
create a ext4 savefile, but i'm not sure how to trigger that, i think
it should use dialog and go to initrd-progs in /sbin

**Comment by wdlkmpx** [comment]
There are various systems designed for these tiny boards or SoCs or whatever they are called. 

Running puppy with a sfs and without pupsave is an improvement.. only read operations on the sd card. 

I think pupmode 13 should indeed default to manual saves, 01micko was right when he wanted to get rid of pup_event_frontend_d, it's all trivial stuff.
I improved it a lot though. I'll add static binaries it's currently a bash script, both behave exactly the same... only PM 13 and auto shutdown stuff. 

So yeah, a frugal install is the only reasonable thing in these tiny boards. 

Ways to achieve this:
- cross-compiling a kernel
- compiling a custom vmlinuz using the same kernelconfig and stuff, using an already working puppy in the target machine?

And a third one:
- the one explained in the first comment

And a fourth one:
- provided by support/kernel_pkg.sh / kernel_pkg_zdrv2initrd.sh

It's is specified in KERNEL_TARBALL, but using a pet, deb,or txz kernel. Although my idea was to make 3builddistro detect if a kernel is being included in rootfs-complete, move the modules the zdrv or fdrv, copy essential modules to the initrd and that should boot in frugal mode.

But i guess i'll delete that code, but i'll experiment one last time, one of these weeks.

**Comment by woodenshoe-wi** [comment]
> Now that i have a raspi i built a raspup, i'm yet to test it.
https://github.com/puppylinux-woof-CE/woof-CE/wiki/Builds

I downloaded it twice and both times it failed checking the md5sum.  I unpacked it onto my extra SD card anyway and it seems to be missing some important files but the main sfs was there so I copied some of the missing files from the build I did in January.

> For some reason some apps don't run (urxvt, retrovol,etc).. not found,
but file identifies them as elf binaries. Must be an issue with the sd
card or ram or aufs?. I don't see aufs errors in dmesg.

I think I figured it out.  If you check the binaries that have the problem, they have been compiled to use /lib/ld-linux.so.3 as interpreter.  Try making a symlink /lib/ld-linux.so.3 -> arm-linux-gnueabihf/ld-2.24.so

As for compiling kernels, it should be possible to cross-compile a kernel on a PC using kernel-kit.  I haven’t compiled one recently, but I did add support for it to kernel-kit using the official Raspberry Pi cross compilers and kernel source.  It was working fine the last time I used it.

**Comment by wdlkmpx** [comment]
gftp is not working right for me. My connectivity with micko's site is
bad, i used filezilla and had to resume like 60 times.

But it's good to know you found the problem, i have to prepare a
special external hard drive (including a 12V whatever) so i can
perform more tests and adapt to my new environment. I'll buy an orange
pi, but i don't know if it'll arrive, and probably add slackwarearm
with cli only stuff, so everything runs on ram.

I have another request, another shortcut

CTRL+T = Window->New Window
ALT+UP_ARROW = Window-> Parent, New Window

I still don't know how to edit rox's shortcuts but i'll learn

**Comment by wdlkmpx** [comment]
Although the best distro for a raspi.. is Alpine Linux, anything you
compile statically is tiny.. musl. No need to compile a special
cross-compiler (initrd-progs). I briefly tested only once in console
mode, but i stopped right away, for some reason. An alpine pup would
be good. And that could happen.

You compile something and probably can even test it in a tablet/phone.

I used to use tablet my some old native android C toolchains, test and
fix my irc bot written in C.. android's bionic is too primitive, but
musl should work better without the need to modify source code. But
it's not idea, it's all flash storage.

One day, i saw my ip was blacklisted by Rizon and depressed me a
little bit, i could have downloaded 4000+ soundtracks from xdcc bots.
Such is life.

**Comment by woodenshoe-wi** [comment]
> CTRL+T = Window->New Window

This one is possible.  Make a file named `.gtkrc.mine` in your home directory and add the line `gtk_can_change_accels=1` then reboot or restart the X server, open the menu to the item you want to set the shortcut for and press the key combination you want.  Pressing “Delete” on an entry that has a key binding will remove the key binding, pressing “Delete” on an entry that does not have a key binding will set the key binding to “Delete”.

Shortcuts involving number keys won’t work because of the “selection group” feature of ROX filer, and shortcuts involving the arrow keys don’t seem to work because of the code that lets you navigate the filer window with the arrow keys.  It might be possible to make the “selection group” feature optional and fix the arrow key handling to ignore arrow keys combined with modifier keys, but it would take work.  (I’m not planning on doing it.)

**Comment by wdlkmpx** [comment]
It seems that compiling kernels is definitely the easy way, or a custom vmlinuz

I removed the code i added
https://github.com/puppylinux-woof-CE/woof-CE/commit/a15cdff5d0f101c1fc456701d049dc4654db6be7

Before that i had removed 75 lines from the same scripts.

At this point the stuff is quite readable and fairly modular, but another massive cleanup is on the way.

The next raspup.zip will be exciting, i'll bump the version from 7 to 8, i noticed 3builddistro does not create the devx.sfs, i'll look into that, it will be a complete build.

**Comment by wdlkmpx** [comment]
It looks like most pthemes are too heavy for a raspi 1, including "Original Pup" which only uses svg images.

So at least "Original Pup" and "Dark_blue" will use only png images and very minimal jwm stuff, the most lightweight global themes suitable for a raspi 1.

It's easy to change and test theme "modules" by clicking on the "More" buttons, the more i test ptheme, the more sense it makes, but if you don't click on them, you'll get the impression that it only changes global themes without a preview... It's a neat app indeed which replaces many others in a single dialog.
